cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package throttling
+
+import (
+	"math"
+	"time"
+)
+
+type tokenBucket struct {
+	creditsPerSecond float64
+	balance          float64
+	maxBalance       float64
+	lastTick         time.Time
+	timeNow          func() time.Time // For testing
+}
+
+func (t *tokenBucket) Drain(maxDrain float64) float64 {
+	now := t.timeNow()
+	interval := now.Sub(t.lastTick)
+	t.lastTick = now
+	diff := interval.Seconds() * t.creditsPerSecond
+	t.balance = math.Min(t.balance+diff, t.maxBalance)
+	result := math.Min(t.balance, maxDrain)
+	t.balance -= result
+	return result
+}
+
+type client struct {
```

Comment:
this is gonna suck but in the future^TM, we're going to have to support multiple services from the same client. However, given how this is an internal piece of work, I think we can make the changes when required.

Reply:
So client here means client to Jaeger agent from within service. How can there be more than one service/application running a Jaeger client?